src/mito2/src/sst/parquet/row_group.rs

```diff
@@ -279,31 +647,74 @@ impl<'a> InMemoryRowGroup<'a> {
 
     /// Try to fetch data from the memory cache or the WriteCache,
     /// if not in WriteCache, fetch data from object store directly.
-    async fn fetch_bytes(&self, ranges: &[Range<u64>]) -> Result<Vec<Bytes>> {
+    async fn fetch_bytes(
+        &self,
+        ranges: &[Range<u64>],
+        metrics: Option<&ParquetFetchMetrics>,
+    ) -> Result<Vec<Bytes>> {
         // Now fetch page timer includes the whole time to read pages.
         let _timer = READ_STAGE_FETCH_PAGES.start_timer();
+
         let page_key = PageKey::new(self.file_id, self.row_group_idx, ranges.to_vec());
         if let Some(pages) = self.cache_strategy.get_pages(&page_key) {
+            if let Some(metrics) = metrics {
+                metrics.inc_page_cache_hit();
+                metrics.add_pages_to_fetch_mem(ranges.len());
+                let total_size: u64 = ranges.iter().map(|r| r.end - r.start).sum();
+                metrics.add_page_size_to_fetch_mem(total_size);
+            }
             return Ok(pages.compressed.clone());
         }
 
+        // Calculate total range size for metrics.
+        let (total_range_size, unaligned_size) = compute_total_range_size(ranges);
+
         let key = IndexKey::new(self.region_id, self.file_id, FileType::Parquet);
-        let pages = match self.fetch_ranges_from_write_cache(key, ranges).await {
-            Some(data) => data,
+        let fetch_write_cache_start = metrics.map(|_| std::time::Instant::now());
+        let write_cache_result = self.fetch_ranges_from_write_cache(key, ranges).await;
+        let pages = match write_cache_result {
+            Some(data) => {
+                if let Some(metrics) = metrics {
+                    let elapsed = fetch_write_cache_start
+                        .map(|start| start.elapsed().as_micros() as u64)
+                        .unwrap_or_default();
+                    metrics.add_write_cache_fetch_elapsed(elapsed);
+                    metrics.inc_write_cache_hit();
+                    metrics.add_pages_to_fetch_write_cache(ranges.len());
+                    metrics.add_page_size_to_fetch_write_cache(unaligned_size);
```

Comment:
do we need to use `aligned_size` to reflect the real IO metric?

Reply:
The aligned size is the size of the referenced buffer, which isn't the actual size to read.